File: voxelMain.py
"""
Project 5
ME C201: Modeling and Simulation of Advanced Manufacturing Processes
Spring Semester 2021
"""

import numpy as np
import matplotlib.pyplot as plt
import scipy.sparse
import scipy.sparse.linalg
from mpl_toolkits.mplot3d import Axes3D
from matplotlib.animation import ArtistAnimation
import matplotlib.cm as cm
import pandas as pd
import os

# Constants                                    |UNITS    | Description
K = 135  # | W/m-K   | Thermal conductivity
# conductivity
C = 450  # | J/kgK   | heat capacity
rho_0 = 6500  # | kg/m^3  | density
theta_w = 300  # |   K     | temp at the wall
theta_0 = 300  # |   K     | initial temp
I_0 = 2e7  # |  W/m^2  | Initial laser intensity
a = 25  # |  1 / m  | absorption constant
tf = 2  # |   s     | final simulated time
dt = 1e-4  # |   s     | time step size
L = 0.05  # |   mm    | side length of cube

# load data from c++ generation of particles
df = pd.read_csv("density.dat", sep=' ')
data = np.asarray(df.values)
xPos = data[:, 0]
yPos = data[:, 1]
zPos = data[:, 2]
density = data[:, 3]
heatCap = data[:, 4]
thermCond = data[:, 5]

def intensity(zz):
    intense = I_0 * np.exp(-a * (L - zz))
    return intense


def laser(zz):
    energy_output = a * I_0 * np.exp(-a * (L - zz))
    return energy_output


def temp_profile(N):
    # global K, C, rho_0, theta_w, theta_0, I_0, a, tf, dt, L
    h = L / (N - 1)
    print('... Computing A ...')

    # compute coordinates for every node
    coord_cube = np.zeros((N ** 3, 4))  # array containing: Node #, x, y, z
    x = y = z = np.arange(0, L + h, h)
    counter = 0
    for k in range(len(z)):
        for j in range(len(y)):
            for i in range(len(x)):
                coord_cube[counter, 0] = counter
                coord_cube[counter, 1], coord_cube[counter, 2], coord_cube[counter, 3] = x[i], y[j], z[k]
                counter = counter + 1

    top_boundary = []
    top_boundary_h = []
    bottom_boundary = []
    bottom_boundary_h = []
    wall_boundary = []
    counter = 0
    for i in range(N):
        for j in range(N):
            for k in range(N):
                if i == 0:
                    bottom_boundary.append(counter)
                    bottom_boundary_h.append(counter + N ** 2)
                if i == N - 1:
                    top_boundary.append(counter)
                    top_boundary_h.append(counter - N ** 2)
                if j == 0 or j == N - 1 or k == 0 or k == N - 1:
                    wall_boundary.append(counter)
                counter += 1

    # Construct A matrix => N^3 x N^3 where N=21
    A = np.zeros((N ** 3, N ** 3))
    for i in range(N ** 3):
        # set components of A (six point stencil) for nodes not on the top or bottom face
        if (N ** 2) < i < ((N ** 3) - (N ** 2)):
            A[i, i] = -6
            A[i, i - 1] = A[i, i + 1] = 1  # capture neighboring nodes in x-direction
            A[i, i - N] = A[i, i + N] = 1  # capture neighboring nodes in y-direction
            A[i, i - N ** 2] = A[i, i + N ** 2] = 1  # capture neighboring nodes in z-direction


    # zero out the rows in A that pertain to nodes on any of the boundaries
    # for i in range(len(boundary_node)):
    #     A[boundary_node[i], :] = 0  # captures Dirichlet BC's. Initializes Neumann BCs
    A[wall_boundary, :] = 0
    A = scipy.sparse.csr_matrix(A)

    print('Finished computing A')

    # Compute the I_abs vector
    I_abs = np.zeros(N ** 3)
    for i in range(len(coord_cube)):
        if 0.02 <= coord_cube[i, 1] <= 0.03 and 0.02 <= coord_cube[i, 2] <= 0.03:
            I_abs[(coord_cube[i, 0]).astype(int)] = laser(coord_cube[i, 3])

    # using forward euler, solve differential equation
    theta = np.ones(N ** 3) * theta_0  # compute initial temperature array:
    theta[wall_boundary] = theta_w
    temp = []

    # r_ims = []
    # fig = plt.figure()

    X, Y = np.meshgrid(x, y)
    t = np.arange(0, tf + dt, dt)  # discretize time
    for i in range(len(t)):
        print('average temp: ', np.average(theta))

        # for some reason the following line does not work
        # theta = theta + dt / rho_0 / C * (K / h ** 2 * A @ theta + I_abs)

        Atheta = A @ theta  # but if you precompute this term it will work
        theta = theta + dt / density / heatCap * (thermCond / h ** 2 * Atheta + I_abs)

        # enforce Neumman BCs
        theta[bottom_boundary] = theta[bottom_boundary_h]
        theta[top_boundary] = theta[top_boundary_h]

        # for mp4
        # theta_plot = []
        # for i in range(N ** 3 - N ** 2, N ** 3):
        #     theta_plot.append(theta[i])
        # theta_plot = np.reshape(theta_plot, (N, N))
        # cp = plt.contourf(X, Y, theta_plot, cmap='hot')
        # r_ims.append(cp.collections)

        if i % 100 == 0:
            temp.append(theta)

    # create mp4
    # cbar = plt.colorbar(cp)
    # cbar.set_label(r'$\theta$ (K)')
    # plt.axes().set_aspect('equal')
    # ani = ArtistAnimation(fig, r_ims, interval=1, repeat=True)
    # ani.save('temperature_time.mp4', writer='ffmpeg')

    # plot average of top layer for every time step
    average_top = []
    average_bottom = []
    for i in range(len(temp)):
        average_top.append(np.average(temp[i][top_boundary]))
        average_bottom.append(np.average(temp[i][bottom_boundary]))

    # print('average_top = ', average_top)
    # print('average_bottom = ', average_bottom)

    time_steps = np.arange(0, len(temp), 1) / 1000
    plt.figure()
    plt.plot(time_steps, average_bottom, label='Bottom face')
    plt.plot(time_steps, average_top, label='Top face')
    plt.legend()
    plt.title('Average Temperatures: Top & Bottom Faces')
    plt.ylabel(r'$\theta$ (K)')
    plt.xlabel(r'Time (s)')
    plt.savefig('average.png')

    # plot contour of top layer of cube
    top_cube_temp = []
    for i in range(N ** 3 - N ** 2, N ** 3):
        top_cube_temp.append(temp[-1][i])

    top_temp = np.reshape(top_cube_temp, (N, N))
    plt.figure()
    cp = plt.contourf(X, Y, top_temp, 15, cmap=cm.hot, vmin=np.amin(top_temp), vmax=np.amax(top_temp))
    m = plt.cm.ScalarMappable(cmap=cm.hot)
    m.set_array(top_temp)
    m.set_clim(np.amin(top_temp), np.amax(top_temp))
    plt.ylabel('y(m)')
    plt.xlabel('x(m)')
    cbar = plt.colorbar(m, boundaries=np.linspace(np.amin(top_temp), np.amax(top_temp), 100))
    cbar.set_label("Temperature (K)")

    plt.title(r'$\theta$ (z = 0.05) distribution on XY plane')
    plt.savefig('laser_output_top.png')

    # plot contour of top layer of cube
    bottom_cube_temp = []
    for i in range(0, N ** 2):
        bottom_cube_temp.append(temp[-1][i])

    bottom_temp = np.reshape(bottom_cube_temp, (N, N))


    plt.figure()
    cp = plt.contourf(X, Y, bottom_temp, 15, cmap=cm.hot, vmin=np.amin(top_temp), vmax=np.amax(top_temp))
    m = plt.cm.ScalarMappable(cmap=cm.hot)
    m.set_array(top_temp)
    m.set_clim(np.amin(top_temp), np.amax(top_temp))
    plt.ylabel('y(m)')
    plt.xlabel('x(m)')
    cbar2 = plt.colorbar(m, boundaries=np.linspace(np.amin(top_temp), np.amax(top_temp), 100))
    cbar2.set_label("Temperature (K)")

    plt.title(r'$\theta$ (z = 0.00) distribution on XY plane')
    plt.savefig('laser_output_bottom.png')

    return temp, coord_cube


tempTotal, cubeCoord = temp_profile(31)

# #save voxels - full grid density comparison
# if os.path.exists("stateTemp.dat"):
#     os.remove("stateTemp.dat")
#     # Print the statement once
#     # the file is deleted
#     print("Previous file deleted.")
save_path1 = "/Users/bhopro/Desktop/Berkeley/MSOL/Projects/Voxel/stateTemp.dat"

f = open(save_path1, "w+")
f.write('X Y Z T \n')
# for i in range(len(tempTotal)):
#     f.write(', T{}'.format(i+1))
# f.write('\n')

for i in range(len(cubeCoord[:, 0])):
    f.write("{} {} {} {} \n".format(cubeCoord[i, 1], cubeCoord[i, 2], cubeCoord[i, 3], tempTotal[-1][i]))




f.close()

File: voxelMain.cpp
//
// Voxel code CPP
// Created by Brian Howell on 8/17/21.
//


#include <iostream>
#include <cmath>
#include <fstream>
#include <chrono>
#include <random>
#include <vector>
#include <algorithm>

// particle parameters
const float THETAW = 300.;                              // |    K    |  temperature at the wall
const float THETA0 = 300.;                              // |    K    |  initial temperature
const double I0 = 10.0; //2. * pow(10, 7);        // |  W/m^2  |  initial laser intensity
const float ABSORB = 25.0;                              // |   1/m   |  absorption constant
const float LENGTH = 0.05;                              // |    m    |  sample length
const int NODE = 21;                                    // |   ___   |  number of nodes
const double INTTHICK = 1.0;                            // |   ---   |  interfacial thickness parameter

// material parameters
const int KP = 903; //85;                               // |  W/m-K  |  thermal conductivity of particle
const int KM = 1;                                       // |  W/m-K  |  thermal conductivity of material
const int CP = 903;                                     // | J/kg-K  |  heat capacity of particle
const int CM = 156;                                     // | J/kg-K  |  heat capacity
const int RHO0P = 2700;                                 // | kg/m^3  |  initial particle density
const int RHO0M = 1000;                                 // | kg/m^3  |  initial material density
const double VP = 0.7;                                  // |   ---   |  volume fraction of particles
const double RPART = LENGTH / 10.;                      // |    m    |  radius of the particles

// simulation parameters
const double TF = 2.;                                   // |    s    |  final simulation time
const double DT = pow(10, -4);            // |    s    |  initial time step
int SIZETIME = (int) TF / DT;                           // |   ---   |  size of time array
double H = LENGTH / (NODE - 1);                         // physical discretization length
const int SIZEA3 = (int) pow(NODE, 3);           // try static_cast
const int SIZEA2 = (int) pow(NODE, 2);

// data outputs
std::ofstream printDensity;
std::ofstream printTemp;

// function declarations
void computeCoord(std::vector< std::vector<double> >&);
// computeCoord - compute the coords in the x-y-z directions
// @param vector< vector<double> > - the 2D vector to be filled
// @param double N -  Number of nodes
// @param double L -  Sample length

void computeBoundary(std::vector< std::vector<int> >&);
// boundaryNodes - computes which nodes are on the sides, top and bottom of the cube
// @params vector< vector<double> > - input 2D vector to be filled
// @params double - number of nodes
// @return vector< vector<double> > - vector of vectors containing each side

void computeAsparse(std::vector< std::vector<int> >&,
                    std::vector< std::vector<int> >&);
// computeAsparse - computes the A matrix, in the sparse variety (see comment below for structure)
// @param - modifies 2D vector into a [n x 4] vector

void computeParticles(std::vector< std::vector<double> >&,
                      std::vector<int>&,
                      std::vector<int>&,
                      double[SIZEA3], double[SIZEA3], double[SIZEA3]);
// computeParticles - compute random particles within the medium of material
// @param - modifies 2D vector and filles with particles

void laserProfile(std::vector< std::vector<double> >&, double[SIZEA3]);
// laserProfile: assigns energy values for each node from the laser
// @param - cubeCoord: cube coordinates for the mesh
// @param - laserValues: values of energy for each node

void solutionScheme(std::vector< std::vector<int> >&,
                    std::vector< std::vector<int> >&,
                    std::vector< std::vector<double> >&,
                    std::vector<double>&,
                    double [SIZEA3], double [SIZEA3],
                    double [SIZEA3], double [SIZEA3]);
// solutionScheme - iteratitive solver in time and space
// @param - ASparse: FDM mesh matrix
// @param - bNodes: nodes of the boundaries
// @param - temperature: solution vector for every time step
// @param - theta: solution vector for a single time step
// @param - density: array for density of each node
// @param - heatCap: array for heat capacity of each node
// @param - thermCond: array for thermal conductivity
// @param - laserValues: energy input from laser at each node

void temp2file(std::vector< std::vector<double> >&,
               std::vector< std::vector<double> >&);
// temp2file - write temperature results from all timesteps to file
// @param - cube coordinates [nParticles][node, x, y, z]
// @param - temperature [timesteps x 1][nParticlesTemp]

void lastTemp2file(std::vector< std::vector<double> >&,
                   std::vector<double>&);

void density2file(std::vector< std::vector<double> >&,
                  double density[SIZEA3], double[SIZEA3],
                  double[SIZEA3]);
// write2file - write cube coordinates and density to file for plotting
// @param - cube coordinates [nParticles x 4] -> [node, x, y, z]
// @param - density [nParticles x 1]

void uniqueVec(std::vector<int> &);
// uniqueVec - modifies input vector to only contain unique indices
// @param - vec: vector to be passed in

void print2dVecInt(std::vector< std::vector<int> >&);
// print2dVec - prints 2D vector
// @param - vec: vector to be passed in

void print2dVecDouble(std::vector< std::vector<double> >&);
// print2dVec - prints 2D vector
// @param - vec: vector to be passed in

int main(){
    std::cout << "total timesteps: " << SIZETIME << std::endl;
    std::cout << "final time " << TF << "s" << std::endl;
    // initialize time
    double time[SIZETIME];
    double counter = 0.0;
    for (int i = 0; i < SIZETIME; i++) {time[i] = counter; counter += DT;}

    // initialize cubeCoord, bNode, and A matrix vectors
    std::vector< std::vector<double> > cubeCoord;
    std::vector< std::vector<int> > bNodes;
    std::vector< std::vector<int> > ASparse;

    // vectors and arrays for particle generation
    std::vector<int> particlesInd;                               // vector holding total indices for each particle
    std::vector<int> particlesInterInd;                          // interfacial distance indices

    // initialize material properties for each node
    double density[SIZEA3];
    double heatCap[SIZEA3];
    double thermCond[SIZEA3];
    std::fill_n(density, SIZEA3, RHO0M);
    std::fill_n(heatCap, SIZEA3, CM);
    std::fill_n(thermCond, SIZEA3, KM);

    // initialize temperature
    std::vector< std::vector<double> > temperature;             // store temperature all time steps
    std::vector<double> theta(SIZEA3, THETA0);                  // store temperature for individual time steps

    // compute necessary matrices
    computeCoord(cubeCoord);                                // compute the x-y-z coordinates
    computeBoundary(bNodes);                                // find the boundary nodes
    computeAsparse(ASparse, bNodes);                    // compute FDM mesh A matrix



    // compute laser profile
    double laserValues[SIZEA3];
    std::fill_n(laserValues, SIZEA3, 0.);
    laserProfile(cubeCoord, laserValues);


    // compute the random particles embedded in the material
    computeParticles(cubeCoord, particlesInd, particlesInterInd,
                     density, heatCap, thermCond);


    solutionScheme(ASparse, bNodes, temperature, theta,
                   density, heatCap, thermCond, laserValues);

    density2file(cubeCoord, density, heatCap, thermCond); // write initial density result to output file
    lastTemp2file(cubeCoord, theta);

    return 0;
}

// Function definitions
void computeCoord(std::vector< std::vector<double> > &cubeCoord){
    /* structure cubeCoord:
     * cubeCoord = [node, x, y, z]
     */

    // required variables
    double coordDist = 0;               // distance incrementer
    std::vector<double> x;              // node numbering in x direction
    std::vector<double> y;              // node numbering in y direction
    std::vector<double> z;              // node numbering in z direction
    std::vector<double> xyz;            // node numbering in xyz direction

    std::cout << "--- Initializing computation --- " << std::endl;
    std::cout << "--- Constructing cubeCoord array ---" << std::endl;
    auto start = std::chrono::high_resolution_clock::now();

    // populate vectors x-y-z
    for (int i = 0; i < NODE; i +=1 ){
//        xyz.push_back(coordDist);
        x.push_back(coordDist);
        y.push_back(coordDist);
        z.push_back(coordDist);
        coordDist += H;
    }

    // populate array coordCube with x-y-z directions
    int counter = 0;
    for (int i=0; i<NODE; i++){
        for (int j=0; j<NODE; j++){
            std::vector<double> temp;
            for (int k=0; k<NODE; k++){
                temp.push_back(counter);
                temp.push_back(x[k]);
                temp.push_back(y[j]);
                temp.push_back(z[i]);
//                std::cout << xyz[k] << " " << xyz[j]  << " " << xyz[i] << std::endl;
                counter++;
                cubeCoord.push_back(temp);
                temp.clear();
            }

        }
    }

    auto stop = std::chrono::high_resolution_clock::now();
    auto duration = (std::chrono::duration_cast<std::chrono::microseconds>(stop - start)).count() / 1e6;
    std::cout << "computational time: " << duration << "s\n" << std::endl;

}

void computeBoundary(std::vector< std::vector<int> >& bNodes){
    std::cout << "--- Initializing computation --- " << std::endl;
    std::cout << "--- Constructing bNodes array ---" << std::endl;
    auto start = std::chrono::high_resolution_clock::now();

    std::vector<int> topBoundary;
    std::vector<int> topBoundary_h;
    std::vector<int> bottomBoundary;
    std::vector<int> bottomBoundary_h;
    std::vector<int> wallBoundary;

    int counter = 0;
    for (int i = 0; i < NODE; i++){
        for (int j = 0; j < NODE; j++){
            for (int k = 0; k < NODE; k++){
                if (i == 0){
                    // log the two boundary nodes
                    bottomBoundary.push_back(counter);              // boundary node
                    bottomBoundary_h.push_back(counter + SIZEA2);   // adjacent boundary node
                }
                if (i == NODE - 1){
                    topBoundary.push_back(counter);                 // boundary node
                    topBoundary_h.push_back(counter - SIZEA2);      // adjacent boundary node
                }
                if (j == 0 or j == NODE - 1 or k == 0 or k == NODE - 1){
                    wallBoundary.push_back(counter);
                }
                counter++;
            }
        }
    }
    bNodes.push_back(topBoundary);
    bNodes.push_back(topBoundary_h);
    bNodes.push_back(bottomBoundary);
    bNodes.push_back(bottomBoundary_h);
    bNodes.push_back(wallBoundary);

    auto stop = std::chrono::high_resolution_clock::now();
    auto duration = (std::chrono::duration_cast<std::chrono::microseconds>(stop - start)).count() / 1e6;
    std::cout << "compute time boundaries: " << duration << "s\n" << std::endl;
}

void computeAsparse(std::vector< std::vector<int> >& ASparse,
                    std::vector< std::vector<int> >& bNodes){
    /* Build a sparse A matrix
     *  A = [N^3, 4]
     *  [node, i, j, value]
     */
    std::cout << "--- Initializing computation --- " << std::endl;
    std::cout << "--- Constructing A sparse array ---" << std::endl;
    auto start = std::chrono::high_resolution_clock::now();

    // Loop through every single node
    for (int i = 0; i < SIZEA3; i++){
        // Loop through only nodes that
        // are not on the bottom or top faces
        if (SIZEA2 < i and i < SIZEA3) {
            std::vector<int> tempVec;
            for (int j = 0; j < SIZEA3; j++) {
                if ((j == i - SIZEA2) or (j == i - NODE) or (j == i - 1)) {
                    tempVec.push_back(i);
                    tempVec.push_back(j);
                    tempVec.push_back(1);
                } else if (j == i) {
                    tempVec.push_back(i);
                    tempVec.push_back(j);
                    tempVec.push_back(-6);
                } else if ((j == i + 1) or (j == i + NODE) or (j == i + SIZEA2)) {
                    tempVec.push_back(i);
                    tempVec.push_back(j);
                    tempVec.push_back(1);
                } else {
                    continue;
                }
                ASparse.push_back(tempVec);
                tempVec.clear();
            }
        }else{
            continue;
        }
    }

    /*
     * enforce dirichlet boundary conditions along the walls
     * loop through boundary wall nodes bNodes[4][i] and set the value of
     * ASparse[node][3] equal to zero
     */
    for (int i = 0; i < bNodes[4].size(); i++){
        for(int j = 0; j < ASparse.size(); j++) {
            if (ASparse[j][0] == bNodes[4][i]) {
                ASparse[j][2] = 0;
            }
        }
    }

    auto stop = std::chrono::high_resolution_clock::now();
    auto duration = (std::chrono::duration_cast<std::chrono::microseconds>(stop - start)).count() / 1e6;
    std::cout << "compute time A matrix: " << duration << "s" << std::endl;
}

void computeParticles(std::vector< std::vector<double> >& cubeCoord,
                      std::vector<int> &particlesInd,
                      std::vector<int> &particlesInterInd,
                      double density[SIZEA3],
                      double heatCap[SIZEA3],
                      double thermCond[SIZEA3]){

    int nParticleNode = std::round(SIZEA3 * VP);     // total number of host nodes for particles
    double partDist, nodeParticle, randLoc, testVar;

    int counter1 = 0;
    while ((particlesInd.size() < nParticleNode) and (counter1 < 10000)){

        // choose a random node to generate particle
        // https://stackoverflow.com/questions/19665818/generate-random-numbers-using-c11-random-library
        std::random_device rd{};
//        std::mt19937 gen{rd()};         // random seed
        std::mt19937 gen{1};            // set seed
        std::uniform_real_distribution<> dist{0, 1};
        randLoc = dist(gen);          // ensure diameter is positive

        // Generate random seed location for particle
        std::cout << "\n-- Generating particle " << counter1 + 1 << std::endl;
        nodeParticle = ceil(SIZEA3 * randLoc);

        // find nodes that are within the distance of the seed location
        for (int i = 0; i < SIZEA3; i++){

            // compute distance between generated particle and all other particles
            partDist = sqrt(pow(cubeCoord[i][1] - cubeCoord[nodeParticle][1], 2) +
                            pow(cubeCoord[i][2] - cubeCoord[nodeParticle][2], 2) +
                            pow(cubeCoord[i][3] - cubeCoord[nodeParticle][3], 2));

            // determine if distance is within the range of another particle
            if (partDist <= RPART){
                particlesInd.push_back(i);
            }else if (INTTHICK != 0 and partDist <= RPART + INTTHICK * H){
                particlesInterInd.push_back(i);

            }else{
                continue;
            }
        }

        uniqueVec(particlesInd);
        uniqueVec(particlesInterInd);

//        std::cout << "particleInd_.size(): " << particlesInd.size() << std::endl;
//        std::cout << "particleInterInd_.size(): " << particlesInterInd.size() << std::endl;

        // assign interfacial material properties
        for (int i = 0; i < particlesInterInd.size(); i++){
            density[particlesInterInd[i]] = (RHO0M + RHO0P) / 2.;
            heatCap[particlesInterInd[i]] = (CP + CM) / 2.;
            thermCond[particlesInterInd[i]] = (KP + KM) / 2.;
        }

        // assign particle material properties
        for (int i = 0; i < particlesInd.size(); i++){
            density[particlesInd[i]] = RHO0P;
            heatCap[particlesInd[i]] = CP;
            thermCond[particlesInd[i]] = KP;
        }
        counter1++;
    }
}

void laserProfile(std::vector< std::vector<double> >& cubeCoord,
                  double laserValues[SIZEA3]){
    for (int i = 0; i < SIZEA3; i++){
//        std::cout << "x: " << cubeCoord[i][1] << " y: " << cubeCoord[i][2] << std::endl;
        if (0.0199 <= cubeCoord[i][1] and cubeCoord[i][1] <= 0.0301 and 0.0199 <= cubeCoord[i][2] and cubeCoord[i][2] <= 0.0301){
            laserValues[i] = ABSORB * I0 * exp(-ABSORB * (LENGTH - cubeCoord[i][3]));
        }
    }
}

void solutionScheme(std::vector< std::vector<int> >& ASparse,
                    std::vector< std::vector<int> >& bNodes,
                    std::vector< std::vector<double> >& temperature,
                    std::vector<double>& theta,
                    double density[SIZEA3], double heatCap[SIZEA3],
                    double thermCond[SIZEA3], double laserValues[SIZEA3]){
    // start timer
    auto start = std::chrono::high_resolution_clock::now();

    // begin time stepping
    double prefix1, prefix2, averageTemp;
    int row, col, val;
    int nnz = ASparse.size();

    for (int t = 0; t < SIZETIME; t++){
        // std::cout output information
        averageTemp = std::accumulate(theta.begin(), theta.end(), 0.0) / theta.size();
//        std::cout << "time: " << t + 1 << " / " << SIZETIME;
//        std::cout << " -> average temperature: " << averageTemp << std::endl;

        // coordinate-wise sparse-matrix-vector multiplication
        // http://www.mathcs.emory.edu/~cheung/Courses/561/Syllabus/3-C/sparse.html
        double AtimesTheta[SIZEA3];
        std::fill_n(AtimesTheta, SIZEA3, 0);
        for (int i = 0; i < nnz; i++){
            row = ASparse[i][0];
            col = ASparse[i][1];
            val = ASparse[i][2];
            AtimesTheta[row] = AtimesTheta[row] + val * theta[col];
        }

        // using solve for temperatures at the next time step
        for (int j = 0; j < SIZEA3; j++){
            theta[j] = theta[j] + DT / density[j] / heatCap[j] *
                    (thermCond[j] / pow(H, 2.0) * AtimesTheta[j] + laserValues[j]);
        }

        // enforce Neumann boundary conditions on the top boundary
        for (int k = 0; k < bNodes[0].size(); k++){
            theta[bNodes[0][k]] = theta[bNodes[1][k]];
        }

        // enforce Neumann boundary conditions on the bottom boundary
        for (int l = 0; l < bNodes[2].size(); l++){
            theta[bNodes[2][l]] = theta[bNodes[3][l]];
        }

        // store temperature results
        if (t % 100 == 0){
            temperature.push_back(theta);
        }else{
            continue;
        }
    }

    std::cout << " -> average temperature: " << averageTemp << std::endl;

    auto stop = std::chrono::high_resolution_clock::now();
    auto duration = (std::chrono::duration_cast<std::chrono::microseconds>(stop - start)).count() / 1e6;
    std::cout << "simulation time: " << duration << "s" << std::endl;
}

void temp2file(std::vector< std::vector<double> >& cubeCoord,
               std::vector< std::vector<double> >& temperature){

    printTemp.open("/Users/bhopro/Desktop/Berkeley/MSOL/Projects/Voxel/stateTemp.dat");
    printTemp << "X Y Z ";
    for (int i = 0; i < temperature.size(); i++){printTemp << "T" << i + 1 << " ";}
    printTemp << "\n";

    // loop through timesteps
    for (int i = 0; i < cubeCoord.size(); i++){
        printTemp << cubeCoord[i][1] << " " << cubeCoord[i][2] << " " << cubeCoord[i][3] << " ";
        for (int j = 0; j < temperature.size(); j++){
            printTemp << temperature[i][j] << " ";
        }
        printTemp << std::endl;
    }
}

void lastTemp2file(std::vector< std::vector<double> >& cubeCoord,
                   std::vector<double>& temperature){
    // write to file
    printTemp.open("/Users/bhopro/Desktop/Berkeley/MSOL/Projects/Voxel/lastTemp.dat");
    printTemp << "X Y Z T" << std::endl;

    for (int i = 0; i < SIZEA3; i++){
        printTemp << cubeCoord[i][1] << " " << cubeCoord[i][2] << " " << cubeCoord[i][3]
                     << " " << temperature[i] << std::endl;
    }
    printTemp.close();
}
void density2file(std::vector< std::vector<double> > &cubeCoord,
                double density[SIZEA3], double heatCap[SIZEA3],
                double thermCond[SIZEA3]){

    // write to file
    printDensity.open("/Users/bhopro/Desktop/Berkeley/MSOL/Projects/Voxel/density.dat");
    printDensity << "X Y Z D C K" << std::endl;

    for (int i = 0; i < SIZEA3; i++){
        printDensity << cubeCoord[i][1] << " " << cubeCoord[i][2] << " " << cubeCoord[i][3]
                   << " " << density[i] << " " << heatCap[i] << " " << thermCond[i] << std::endl;
    }
    printDensity.close();
}

void uniqueVec(std::vector<int> &vec){
    // return only unique nodes in particleInd
    // https://www.geeksforgeeks.org/stdunique-in-cpp/

//    std::cout << "before: " << std::endl;
//    for (int i = 0; i < vec.size(); i++){
//        std::cout << vec[i] << " ";
//    }
//    std::cout << std::endl;

    // begin removing duplicate indices
    std::vector<int>::iterator ip;
    ip = std::unique(vec.begin(), vec.begin() + vec.size());
    vec.resize(std::distance(vec.begin(), ip));

//    std::cout << "after: " << std::endl;
//    for (ip = vec.begin(); ip != vec.end(); ++ip) {
//        std::cout << *ip << " ";
//    }
//    std::cout << std::endl;
}

void print2dVecDouble(std::vector< std::vector<double> >& vec){
    for (int i = 0; i < vec.size(); i++){
        for (int j = 0; j < vec[i].size(); j++){
            std::cout << vec[i][j] << " ";
        }
        std::cout << std::endl;
    }
}

void print2dVecInt(std::vector< std::vector<int> >& vec){
    for (int i=0; i<vec.size(); i++){
        for (int j=0; j<vec[i].size(); j++){
            std::cout << vec[i][j] << " ";
        }
        std::cout << std::endl;
    }
}





